### PR TITLE
:running: Install binaries required for running tests

### LINKED ISSUE #887: Add missing binaries for envtest
For the envtest to usable, the ```etcd, kubectl and kube-apiserver``` binaries are required. As we have it now, none of the make targets install the binaries.
Due to this issue, the current test suit is skipped and none of the tests could run, preventing creation of new controllers’ tests.

Therefore, adding the following is beneficial.

1. Add a make target to install the missing binaries
2. Change the ```test``` target taking these binaries into account
3. Add a ```test-cover``` target

/kind feature

### PR DESCRIPTION
**Pull request by Xenwar**
Signed-off-by: Anwar Hassen <anwar.hassen@est.tech>

Install binaries required for running tests and enables skipped test

Binary installer taken https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/Makefile

Fixes #887

### CODE REVIEW

**Comment by jichenjc** [comment]
@Xenwar do you mind create an issue to describe the gaps/issues that this PR addressed? it will be easier to know from there.. thanks

**Comment by Xenwar** [comment]
> @Xenwar do you mind create an issue to describe the gaps/issues that this PR addressed? it will be easier to know from there.. thanks

Added a new issue and liked it. #887

File: Makefile [hunk 1]
```diff
@@ -106,7 +106,13 @@ $(ARTIFACTS):
 
 .PHONY: test
 test: ## Run tests
-	go test -v ./...
+	source ./scripts/fetch_ext_bins.sh; fetch_tools; setup_envs; go test -v ./...
+
+.PHONY: test-cover
+test-cover: ## Run tests with code coverage and code generate  reports
```
Comment (jichenjc) [inline Makefile:112]: thanks for submit this,

I run locally and it works well , one nit is do we want to rename this to something like capi_coverage to make it more specifi?

File: Makefile [hunk 2]
```diff
@@ -106,7 +106,13 @@ $(ARTIFACTS):
 
 .PHONY: test
 test: ## Run tests
-	go test -v ./...
+	source ./scripts/fetch_ext_bins.sh; fetch_tools; setup_envs; go test -v ./...
```
Comment (jichenjc) [inline Makefile:109]: do we want to run test also include cover or just the test itself?

File: Makefile (hunk 1 above)
Comment (sbueringer) [reply Makefile:112]: I would really like to keep this consistent with the main repo. I don't see a need to prefix this which anything to be honest :)

File: Makefile (hunk 2 above)
Comment (sbueringer) [reply Makefile:109]: Just the test, please compare with the main repo. I wouldn't want to do it differently here if we don't have to

Comment (sbueringer) [inline Makefile:109]: @Xenwar Can you maybe drop the -v in the test target and add a test-verbose (analog to the cluster-api repo)

File: Makefile [hunk 3]
```diff
@@ -106,7 +106,13 @@ $(ARTIFACTS):
...
+
+.PHONY: test-cover
+test-cover: ## Run tests with code coverage and code generate  reports
+	source ./scripts/fetch_ext_bins.sh; fetch_tools; setup_envs; mkdir -p /tmp/out; go test -v -coverprofile=/tmp/out/coverage.out ./... $(TEST_ARGS)
```
Comment (sbueringer) [inline Makefile:113]: Can we call the test target from above here? (please do it the same way as in the main repo, e.g. this also means using out/coverage.out instead the tmp folder)

File: scripts/fetch_ext_bins.sh [hunk 4]
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env bash
+# Copyright 2018 The Kubernetes Authors.
```
Comment (sbueringer) [inline scripts/fetch_ext_bins.sh:2]: I'm not sure where this is from but I would prefer the version from the cluster-api repo. In this script here we delete a folder when SKIP_FETCH_TOOLS is set, which is a bit strange, as imho it shouldn't do anything.

Overall we try to stay consistent with the main cluster-api repo in CAPO to reduce maintenance effort on our side.

File: Makefile (hunk 1 above)
Comment (Xenwar) [reply Makefile:112]: I agree with not prefixing it.

File: Makefile (hunk 2 above)
Comment (Xenwar) [reply Makefile:109]: Thanks. Removed

File: Makefile (hunk 3 above)
Comment (Xenwar) [reply Makefile:113]: thanks. done

File: Makefile (hunk 2 above)
Comment (Xenwar) [reply Makefile:109]: for the cover, there is another target.

File: scripts/fetch_ext_bins.sh (hunk 4 above)
Comment (Xenwar) [reply scripts/fetch_ext_bins.sh:2]: The capi repo does not have the script. But, the ones in Amazon, which which I took this script and the one in GCP use the same approach. 
https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/main/scripts/fetch_ext_bins.sh

**Comment by Xenwar** [comment]
@jichenjc @sbueringer
Sorry for the late reply, could not get time due to higher priority PRs. 
I have made the relevant changes as suggested.

**Comment by jichenjc** [comment]
#938 merged, hopefully the gate will be better now
/test pull-cluster-api-provider-openstack-e2e-test

**Comment by jichenjc** [comment]
/test pull-cluster-api-provider-openstack-e2e-test

#938 merged, let's try again

**Pull request closed without merging**